Fail to run GC job

1. Install 2.1 build with local file system.
2. Replicate some images to projects 
3. Delete images in these projects
4. Change the GC_TIME_WINDOW_HOURS=0 and restart Harbor.
5. Run GC manually. the GC job failed with the following logs:  all image files such as blobs and repository remain in the /data/registry.
```
2020-08-21T10:24:04Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:257]: delete the manifest with registry v2 API: library/haproxy, application/vnd.docker.distribution.manifest.v2+json, sha256:8a0fdda4d13959c085d949fae822dae5fd2d384208a6a9600ebb46ea3baaff23
2020-08-21T10:24:04Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:269]: delete manifest from storage: sha256:8a0fdda4d13959c085d949fae822dae5fd2d384208a6a9600ebb46ea3baaff23
2020-08-21T10:24:04Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:281]: delete artifact trash record from database: 2, library/haproxy, sha256:8a0fdda4d13959c085d949fae822dae5fd2d384208a6a9600ebb46ea3baaff23
2020-08-21T10:24:04Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:293]: delete blob from storage: sha256:8a0fdda4d13959c085d949fae822dae5fd2d384208a6a9600ebb46ea3baaff23
2020-08-21T10:24:04Z [ERROR] [/jobservice/job/impl/gc/garbage_collection.go:163]: failed to execute GC job at sweep phase, error: failed to delete blob from storage: sha256:8a0fdda4d13959c085d949fae822dae5fd2d384208a6a9600ebb46ea3baaff23, deletefailed: http status code: 500, body: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
```

Hello,

I saw that the problem was also refered in #13589, but this issue seems more specific about this problem.

I have the same issue with nfs v3 and Harbor v2.1.2, so I tested with 4.0 and 4.1, but the issue remains :( 

Here are the nfs parameters :

```
(rw,relatime,vers=4.1,rsize=65536,wsize=65536,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=REDACTED,local_lock=none,addr=REDACTED)
```

There are the following logs in the registryctl :

```
2021-02-19T15:51:00.188427+01:00 registryctl 369960bb3a57 time="2021-02-19T14:51:00Z" level=info msg="deleting manifest: /docker/registry/v2/repositories/stresstest-trivy/logstash/_manifests/revisions/sha256/6a54dc5e09dfdf25d0058cd878fd9b
483b25a162cfa89d6c4071fcf9b26610c8" go.version=go1.14.7
2021-02-19T15:51:00.189254+01:00 registryctl 369960bb3a57 10.88.0.107 - - [19/Feb/2021:14:51:00 +0000] "DELETE /api/registry/stresstest-trivy/logstash/manifests/sha256:6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8 HTTP/
1.1" 200 0
2021-02-19T15:51:00.190614+01:00 registryctl 369960bb3a57 time="2021-02-19T14:51:00Z" level=info msg="Deleting blob: /docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8" go.version=go1.14.7
2021-02-19T15:51:00.193781+01:00 registryctl 369960bb3a57 2021-02-19T14:51:00Z [INFO] [/registryctl/api/registry/blob/blob.go:44]: failed to remove blob: sha256:6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8, with error:filesystem: unlinkat /storage/docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8: directory not empty                                                                                        
2021-02-19T15:51:00.193991+01:00 registryctl 369960bb3a57 2021-02-19T14:51:00Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"unknown: filesystem: unlinkat /storage/docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8: directory not empty"}]}
```

Indeed there is a .nfs000* file

```
# ls -al /harbor-data/registry/docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8/
total 16
drwxr-x---  2 10000 10000 4096 Feb 19 15:51 .
drwxr-x--- 12 10000 10000 4096 Feb 19 14:57 ..
-rw-r-----  1 10000 10000 5106 Feb 19 14:57 .nfs000000000000480300000001
```

I can post the content of the .nfs000000000000480300000001 file if needed.

A lsof shows that the .nfs000* file is opened by the registry-photon container, I'm not sure that it means that the deleted file was also opened by this container.

```
# lsof /harbor-data/registry/docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8/.nfs000000000000480300000001
lsof: no pwd entry for UID 10000
COMMAND     PID     USER   FD   TYPE DEVICE SIZE/OFF  NODE NAME
lsof: no pwd entry for UID 10000
registry_ 77372    10000   11r   REG   0,56     5106 18435 /harbor-data/registry/docker/registry/v2/blobs/sha256/6a/6a54dc5e09dfdf25d0058cd878fd9b483b25a162cfa89d6c4071fcf9b26610c8/.nfs000000000000480300000001

# ps -ef |grep 77372
10000      77372   77202  0 15:49 ?        00:00:00 /usr/bin/registry_DO_NOT_USE_GC serve /etc/registry/config.yml
```

As I understand it it means that the deleted file is opened by something at the moment it's deleted.
An easy but ugly win would be to allow a directory to be deleted when it only contains .nfs000* files.

Let me know if you need me to test anyhting.

--- a/make/photon/registry/builder
+++ b/make/photon/registry/builder
@@ -20,13 +20,21 @@ cur=$PWD
 
 # the temp folder to store distribution source code...
 TEMP=`mktemp -d ${TMPDIR-/tmp}/distribution.XXXXXX`
-git clone -b $VERSION https://github.com/docker/distribution.git $TEMP
+git clone -b $VERSION https://github.com/distribution/distribution.git $TEMP
 
 # add patch 2879
-echo 'add patch https://github.com/docker/distribution/pull/2879 ...'
+echo 'add patch https://github.com/distribution/distribution/pull/2879 ...'
 cd $TEMP
-wget https://github.com/docker/distribution/pull/2879.patch
+wget https://github.com/distribution/distribution/pull/2879.patch
 git apply 2879.patch
+
+# add patch 3370
+echo 'add patch https://github.com/distribution/distribution/pull/3370 ...'
+cd $TEMP
+wget https://github.com/distribution/distribution/pull/3370.patch
+git apply 3370.patch
+
+# add patch redis
 git apply $cur/redis.patch
 cd $cur
 

